include/deal.II/trilinos/nox.templates.h

```diff
@@ -1089,8 +1092,25 @@ namespace TrilinosWrappers
                 "has been attached to the NOXSolver object."));
 
             Assert(false, ExcNotImplemented());
-            return 1;
+            ret_code = 1;
           }
+
+        // NOX has a recovery feature that is enabled by default. In this case,
+        // if a solve_with_jacobian or a
+        // solve_with_jacobian_and_track_n_linear_iterations function triggers
+        // an exception and therefore call_and_possibly_capture_exception
+        // returns code different from 0, then NOX does not interrupt the
+        // solution process but rather performs a recovery step. To ensure this
+        // feature is available to the user, we need to supress the exception in
+        // this case, since it is exactly that, what NOX expects from our
+        // callbacks.
+        const bool do_rescue =
+          parameters->sublist("Newton").get("Rescue Bad Newton Solve", true);
+
+        if (do_rescue && pending_exception != nullptr)
+          pending_exception = nullptr;
```

Comment:
This isn't right. First, by setting `pending_exception=nullptr`, you are simply abandoning the exception. This is a resource leak. You need to rethrow the exception, catch it, and ignore it if you want to get rid of it. Take a look at how this is done in `kinsol.cc`:
```
    if (pending_exception)
      {
        try
          {
            std::rethrow_exception(pending_exception);
          }
        catch (const RecoverableUserCallbackError &exc)
          {
            pending_exception = nullptr;
            if (status == 0)
              /* just eat the exception */;
            else
              throw;
          }
        catch (...)
          {
            pending_exception = nullptr;
            throw;
          }
      }
```
Second, you only want to do that if the captured exception is of type `RecoverableUserCallbackError`. For every other error, I will still argue that NOX needs to be made to abort its operations and return to user space.

Reply:
Finally, if we do capture exceptions in these lambda functions, we need to deal with this at the point where we call NOX. Again, kinsol is the right place to look: https://github.com/dealii/dealii/blob/master/source/sundials/kinsol.cc#L475-L505

Reply:
> This isn't right. First, by setting `pending_exception=nullptr`, you are simply abandoning the exception. This is a resource leak. You need to rethrow the exception, catch it, and ignore it if you want to get rid of it. Take a look at how this is done in `kinsol.cc`:

> Second, you only want to do that if the captured exception is of type `RecoverableUserCallbackError`. For every other error, I will still argue that NOX needs to be made to abort its operations and return to user space.

Both points sound reasonable, I will update the code accordingly.

> Finally, if we do capture exceptions in these lambda functions, we need to deal with this at the point where we call NOX. Again, kinsol is the right place to look: https://github.com/dealii/dealii/blob/master/source/sundials/kinsol.cc#L475-L505

I am not sure I got it right. In these lines indeed there is an exception handling after calling `KINSol()`, but there is a similar thing happening inside `SUNDIALS::Utilities::call_and_possibly_capture_exception()` for every callback call: https://github.com/dealii/dealii/blob/8ad845fa53f571021c5919b6f0842f7a3cb90d2a/include/deal.II/sundials/utilities.h#L57-L73